JWT 인증 기능 구현

nonoro opened the pull request:
- jwt 를 이용한 사용자 인증 구현

nonoro committed:
feat: 로그인, 회원가입 view, 버튼 생성

nonoro committed:
merge 커밋

nonoro committed:
Merge remote-tracking branch 'origin' into feature/member

nonoro committed:
gradle: security, jwt 의존성 추가

nonoro committed:
yml: jwt 추가

nonoro committed:
feat: jwt 필터 구현

nonoro committed:
test: jwtProvider 단위 테스트

nonoro committed:
불필요한 클래스 삭제

nonoro committed:
refactor: JwtProvider
, JwtProviderTest 리팩토링

nonoro committed:
Merge remote-tracking branch 'origin' into feature/member

nonoro committed:
주석 제거

nonoro committed:
refactor: kotest로 변경 및 예외 분리

nonoro committed:
feat: jwtFilter 구현

nonoro committed:
build: mockk 의존성 추가

nonoro committed:
refactor: JwtFilter permitAll Api 검증 로직 추가

nonoro committed:
test: JwtFilter 테스트 코드 작성

nonoro committed:
test: JwtFilter 테스트 코드 정상 작동한 Case 추가

nooose commented on wolfdesk/modules/base/src/test/kotlin/wolfdesk/base/filter/jwt/JwtFilterTest.kt:
```diff
@@ -64,4 +67,23 @@ class JwtFilterTest : StringSpec({
+
+    "요청과 함께 온 token이 유효할 경우 사용자 인증이 이루어진다" {
+        val validToken = validJwtProvider.generateToken(memberId = memberId)
+        every { request.getHeader(HttpHeaders.AUTHORIZATION) } returns "Bearer $validToken"
+        every { request.remoteAddr } returns "127.0.0.1"
+        every { request.getSession(false) } returns null
```
모킹없이 `MockHttpServletRequest`으로 테스트할 수 있음

nonoro replied:
MockHttpServletRequest, MockHttpServletResponse로 바꿔서 해놈 확인점

nooose commented on wolfdesk/modules/base/src/main/kotlin/wolfdesk/base/security/SecurityConfig.kt:
```diff
@@ -0,0 +1,47 @@
+class SecurityConfig(
+    private val jwtFilter: JwtFilter,
+) {
+    companion object {
+        val PERMITALL_URIS = setOf(
+            "/members/login",
+            "/members/register",
+        )
+    }
```
companion object는 클래스 제일 아래에 배치하는 것이 컨벤션임

nonoro replied:
수정완료

nooose commented on wolfdesk/modules/base/src/main/kotlin/wolfdesk/base/security/SecurityConfig.kt:
```diff
@@ -0,0 +1,47 @@
+                sessionCreationPolicy = SessionCreationPolicy.STATELESS
+            }
+
+            httpBasic { }
```
이건 왜 함?

nonoro replied:
모름 ... 지워버림

nooose commented on wolfdesk/modules/base/src/main/kotlin/wolfdesk/base/filter/jwt/JwtFilter.kt:
```diff
@@ -0,0 +1,60 @@
+    private fun authenticateUser(token: String, request: HttpServletRequest) {
+        val memberId = jwtProvider.extractMemberId(token)
+
+        if (SecurityContextHolder.getContext().authentication == null) {
+            val userDetails = userDetailsService.loadUserByUsername(memberId.toString())
+            val authentication = UsernamePasswordAuthenticationToken(
+                userDetails, null, userDetails.authorities
+            )
+
+            authentication.details = WebAuthenticationDetailsSource().buildDetails(request)
+            SecurityContextHolder.getContext().authentication = authentication
+        }
```
null이 아닐 때 early return으로 변경하기

nonoro replied:
변경하였읍미다

nooose commented on wolfdesk/modules/base/src/main/kotlin/wolfdesk/base/filter/jwt/JwtFilter.kt:
```diff
@@ -0,0 +1,60 @@
+        val memberId = jwtProvider.extractMemberId(token)
+
+        if (SecurityContextHolder.getContext().authentication == null) {
+            val userDetails = userDetailsService.loadUserByUsername(memberId.toString())
+            val authentication = UsernamePasswordAuthenticationToken(
+                userDetails, null, userDetails.authorities
+            )
```
UsernamePasswordAuthenticationToken 대신 AbstractAuthenticationToken 상속받아 인증토큰객체 직접 만들어보기
그리고 커스텀 로그인을 만들어야 하기 때문에 UserDetailsService를 사용할 이유는 없어보임

nonoro replied:
만들었음 ㅎㅎ

nooose commented on wolfdesk/modules/base/src/main/kotlin/wolfdesk/base/filter/jwt/JwtFilter.kt:
```diff
@@ -0,0 +1,60 @@
+        response: HttpServletResponse,
+        filterChain: FilterChain
+    ) {
+        if (shouldSkipFilter(request)) {
+            filterChain.doFilter(request, response)
+            return
+        }
```
OncePerRequestFilter 에서 shouldNotFilter 함수를 제공하고 있음, 요거를 오버라이드해서 사용하면됨

nonoro replied:
완료~!

nonoro committed:
refactor: HttpServletRequest -> MockHttpServletRequest 변경

nonoro committed:
refactor: securityConfig comment 반영

- 불필요한 코드 삭제
- companion object 컨벤션에 따라 클래스 맨 아래로 배치

nonoro committed:
refactor: JwtFilter comment 반영

- 커스텀 인증 객체 구현
- shouldNotFilter 오버라이드

nooose commented on wolfdesk/modules/base/src/main/kotlin/wolfdesk/base/filter/jwt/CustomAuthenticationToken.kt:
```diff
@@ -0,0 +1,18 @@
+import org.springframework.security.core.GrantedAuthority
+
+class CustomAuthenticationToken(
+    private val principal: Any,
```
인증주체를 나중에 꺼내서 쓰게될텐데 Any 타입이면 매번 형변환필요할 수 있음

이것도 하나 만들어서 사용하는 게 좋아보임

nonoro replied:
ㅇㅋㄷㅋ

nooose commented on wolfdesk/modules/base/src/main/kotlin/wolfdesk/base/filter/jwt/CustomAuthenticationToken.kt:
```diff
@@ -0,0 +1,18 @@
+
+class CustomAuthenticationToken(
+    private val principal: Any,
+    private val credentials: Any? = null,
```
credentials 필요하지 않으니 인스턴스 변수로 가지고 있을필요는 없어보임

nonoro replied:
제거하고 오버라이드한 함수에서 null 리턴하도록 변경함

nooose commented on wolfdesk/modules/base/src/main/kotlin/wolfdesk/base/filter/jwt/JwtFilter.kt:
```diff
@@ -15,46 +13,49 @@ import wolfdesk.base.support.bearerToken
         filterChain: FilterChain
     ) {
-        if (shouldSkipFilter(request)) {
+        if (shouldNotFilter(request)) {
```
shouldNotFilter 오버라이드만 해주면 어차피 상위 콜스택에서 호출하기 때문에
추가로 실행할 이유없음

shouldNotFilter에서 걸리면 `doFilterInternal`를 호출하지 않음 OncePerRequestFilter 들어가서 확인ㄱㄱ

nonoro replied:
<img width="1056" alt="image" src="https://github.com/user-attachments/assets/96f354e3-7ab8-4dee-81d8-db6ca109aa70">
확인

nooose commented on wolfdesk/modules/base/src/main/kotlin/wolfdesk/base/filter/jwt/JwtFilter.kt:
```diff
@@ -15,46 +13,49 @@ import wolfdesk.base.support.bearerToken
+            request = request,
+            response = response,
+        )
 
```
이건 뭐하는거?

nonoro replied:
인증객체 만들어서 넣어주는거

nooose commented on wolfdesk/modules/base/src/main/kotlin/wolfdesk/base/filter/jwt/JwtFilter.kt:
```diff
@@ -15,46 +13,49 @@ import wolfdesk.base.support.bearerToken
 
-    private fun isTokenAbsent(token: String?) = token.isNullOrBlank()
+    override fun shouldNotFilter(request: HttpServletRequest): Boolean {
+        return request.requestURI in SecurityConfig.PERMITALL_URIS
```
`PERMIT_ALL_URIS` 이렇게 바꾸는게 좋지않겠음?

nonoro replied:
ㅇㅋㅇㅋ

nooose commented on wolfdesk/modules/base/src/main/kotlin/wolfdesk/base/filter/jwt/JwtFilter.kt:
```diff
@@ -15,46 +13,49 @@ import wolfdesk.base.support.bearerToken
-    private fun authenticateUser(token: String, request: HttpServletRequest) {
-        val memberId = jwtProvider.extractMemberId(token)
+    private fun authenticateUser(token: String, request: HttpServletRequest, response: HttpServletResponse) {
+        if (isTokenInvalid(token)) {
```
isInvalidToken이 더 자연스러워보임

nooose replied:
따라가면 token.isNullOrBlank() 사용하는 것 같은데 
이정도면 isInvalidToken 함수를 사용하지 않아도될것같음

nonoro replied:
변경완료

nooose commented on wolfdesk/modules/base/src/main/kotlin/wolfdesk/base/filter/jwt/JwtFilter.kt:
```diff
@@ -15,46 +13,49 @@ import wolfdesk.base.support.bearerToken
-        val memberId = jwtProvider.extractMemberId(token)
+    private fun authenticateUser(token: String, request: HttpServletRequest, response: HttpServletResponse) {
+        if (isTokenInvalid(token)) {
+            response.sendError(HttpServletResponse.SC_UNAUTHORIZED, "인증 실패 - 토큰이 없습니다")
+            return
```
이렇게 처리하지말고 예외를 던지게 만들고 AuthenticationEntryPoint 같은것을 만들어서 등록하는것이 좋음

nooose commented on wolfdesk/modules/base/src/main/kotlin/wolfdesk/base/filter/jwt/JwtFilter.kt:
```diff
@@ -15,46 +13,49 @@ import wolfdesk.base.support.bearerToken
+
+        // TODO: 향후 권한이 추가되면 Token 에서 꺼내서 CustomAuthenticationToken 의 인자값으로 권한리스트도 넣어줘야 함
+        val authentication = CustomAuthenticationToken(memberId)
+        authentication.details = WebAuthenticationDetailsSource().buildDetails(request)
```
이거 꼭 해야함? 궁금해서 질문임

nonoro replied:
모름 나도 다들 이렇게하길래 이렇게함 ...

nooose commented on wolfdesk/modules/base/src/main/kotlin/wolfdesk/base/filter/jwt/JwtFilter.kt:
```diff
@@ -15,46 +13,49 @@ import wolfdesk.base.support.bearerToken
-            SecurityContextHolder.getContext().authentication = authentication
+        if (SecurityContextHolder.getContext().authentication != null) {
+            return
         }
```
인증토큰을 시큐리티홀더에 넣어주고 다음 필터를 호출해야함
그래야 정상적인 필터체인이 연결됨

그래서 필터에서 return 하면 정상적인 동작이 되지않음

nooose replied:
Stateless 정책이기 때문에 매요청마다 인증을 새로해야함
그러므로 JwtFilter에서 SecurityContextHolder.getContext().authentication을 까보면 항상 null임

JwtFilter가 하는 일은 Jwt 토큰을 얻어 인증객체를 만들고 시큐리티홀더에 담아주는 역할을 해야함

nonoro replied:
그거 저 아래서 하는데 그럼 저 리턴을 걍 없애? 어짜피 실행안되려나?

nooose replied:
불필요한거같음
JWT 추출불가능하면 인증되지 않은 Authentication을 넘기면 좋을 거같음

nonoro replied:
이거 이쪽 로직 오기전에 jwt 추출불가능하면 그냥 예외로 날려버려서 그냥 저부분 없애고 저기 들어온 순간 무조건 jwt 인증 됐다고 보고 인증객체 만들어서 넣고 다음 필터로 넘기는걸로 했음!

nonoro committed:
refactor: comment 반영2

- 커스텀 principal 구현
- shouldNotFilter 오버라이드 하고 불필요하게 사용한 부분 제거
- 용어 수정
- AuthenticationEntryPoint 및 인증 실패 커스텀 예외 구현
- 전반적으로 확인 후 불필요한 코드 제거

nonoro committed:
refactor: comment 반영2

- 불필요한 개행제거 및 코드 수정

nonoro committed:
refactor: 토큰 생성 시 member 정보를 객체로 받도록 변경 및 이에 따른 테스트코드 수정 + securityConfig 의존성 주입 잘못받고있던 부분 수정

nonoro committed:
feat: member 로그인, 회원가입 기능 구현

nonoro committed:
refactor: jwt 관련 리팩토링

nonoro committed:
feat: ApiResponse 규격에 error 추가

nonoro committed:
refactor: 시큐리티 리팩토링

nonoro committed:
feat: 엔티티 스캔에 멤버 도메인 추가

nonoro committed:
refactor: api 수정

nonoro committed:
config: secretKet default 값추가

nooose committed:
Merge branch 'main' into feature/member

nooose committed:
refactor: JWT 의존성 base 모듈로 이동
